### DISCUSSION TITLE
Ukrainian model proposal

### DISCUSSION THREAD

**Discussion by kurnosovv** [explosion/spaCy#10561] [Language Support]
@honnibal @adrianeboyd
I created SpaCy configs for Ukrainian language, and would like to propose to train the model and add them to the official SpaCy models registry.

Training configs are 
- https://github.com/kurnosovv/ukr-spacy/tree/main/spacy_project (CPU model with pretrained word2vec vectors)
- https://github.com/kurnosovv/ukr-spacy/tree/main/spacy_project_trf (transformer-based model).

All training sources are available under MIT license. Training and evaluation data are silver standard datasets for Ukrainian language.

Could you train and add such models to official SpaCy models registry? During my attempts, training stops much earlier than specified num_epochs, which seems lead to lower accuracy than may be achieved in the best case. Maybe you can modify training procedure, or change training config, so that it will lead to higher-accuracy models. I may assist you, and answer any questions.

**Comment by adrianeboyd** [comment]
Thanks, it would be great to be able to add Ukrainian pipelines! Can you provide some more information about the sources/citations for the training data? If you'd prefer to discuss it over email, you can contact me at adriane@explosion.ai.

**Comment by kurnosovv** [comment]
At the moment, gold-standard Ukrainian language datasets are:
- NER: https://github.com/lang-uk/ner-uk
- Morphology and syntax: https://universaldependencies.org/treebanks/uk_iu/index.html
Both of them are under CC BY-NC-SA 4.0 licence.

For creating silver-standard data, the following steps were performed:
- I trained transformer-based models with quite high accuracy using these datasets
- I used a News subset of Ukrainian UberText Corpus (described here https://lang.org.ua/en/corpora/), it contains shuffled sentences extracted from news articles. I took subsamples of ~1M sentences for training data, and ~100K sentences for validation data
- I applied transformer-based models to make predictions for sampled sentences, and saved them in conllu format
- word2vec vectors are taken from https://lang.org.ua/en/models/ (UberCorpus, 300d as is)

The resulting dataset is synthetic data, stored on my Google Drive (I have not posted it anywhere at the moment)

**Reply by adrianeboyd** [reply]
Thanks for the info! For Russian the trained pipelines are trained on similar silver data, so the main concern on our end is having some kind of citation (at least a link) for users who want to know more about the dataset.

If all the text licenses permit it, I wonder if a relatively simple option would be to publish the data as a huggingface dataset. There are some existing CoNLL-U dataset builders, e.g. for:

https://huggingface.co/datasets/universal_dependencies
https://huggingface.co/datasets/universal_dependencies/blob/main/universal_dependencies.py

(Note that this model card desperately needs more information (it's hard-coded to UD r2.7 but you have to look at the code to figure this out), and I haven't actually tested this particular dataset builder script in detail.)

I had a look at the training data and one thing I noticed is that there's not much punctuation. This is probably going to cause problems for the parser and sentence segmentation. This looks like it comes from the source texts, but I'm not sure what's going on with the preprocessing for these text corpora in general for them to end up like this? In particular, I would have expected to find both commas and sentence-final punctuation?

**Reply by kurnosovv** [reply]
I will need some time to fix the dataset and publish it. May I use web dataset (Common Crawl) instead of news dataset, does it make any difference?

**Reply by adrianeboyd** [reply]
I would be worried that Common Crawl is too noisy to provide good texts for this kind of silver annotation since the models are trained on news texts? And the licensing for the texts is a problem if you want to publish it directly.

**Reply by kurnosovv** [reply]
@adrianeboyd I modified the dataset. Instead of Ukrainian UberText Corpus, now the News subset of Leipzig Corpora Collection (https://wortschatz.uni-leipzig.de/en/download/Ukrainian) is used as corpus. The annotation process is the same. 

The data is published as a huggingface dataset:   https://huggingface.co/datasets/ukr-models/Ukr-Synth

**Reply by adrianeboyd** [reply]
Thanks, can you add a license to the model card?

(As a note, the conllu data looks fine, but some of the columns in the dataset look like they might not have been configured correctly: why are `upos` and `xpos` the same data in a different format, the type of `heads`, and the type/values of `deps`.)

**Reply by kurnosovv** [reply]
License information added
